### PR TITLE
Add Storage FIO testcases

### PR DESCRIPTION
**Pull request by smit-gardhariya**
This PR contains generic FIO testcase addition. Based on runbook params, FIO will be launched for storage perf test. We can pass list of params to cover more than one scenario. If nothing is passed, testcase will launch default testcases mentioned in json file

### CODE REVIEW

**Comment by LiliDeng** [comment]
@smit-gardhariya please check perf_premium_datadisks_4k, whether they are duplicated, if not, you can refer this case with different parameters of fio.

File: microsoft/testsuites/performance/storageperf.py [hunk 1]
```diff
@@ -187,6 +194,109 @@ def perf_storage_over_nfs_synthetic_tcp_4k(self, result: TestResult) -> None:
...
+                fio_result[test_name]["metrics"] = {}
+                log.error(e)
+
+            rm.remove_directory(str(path.joinpath(f"{filename}*")), sudo=True)
+
+        log.debug(fio_result)
+
+        # Saving the result
+        report_path = node.get_working_path().joinpath("fio_report")
+        report_file = report_path.joinpath("fio_test_result.json")
```
Comment (squirrelsc) [inline microsoft/testsuites/performance/storageperf.py]: Don't use a file to dump the test results, it's not extendable. Use standard way, send sub tests messages to dump the results.

**Comment by smit-gardhariya** [comment]
> @microsoft-github-policy-service agree company="Microsoft"

File: microsoft/testsuites/performance/storageperf.py [hunk 2]
```diff
@@ -187,6 +194,109 @@ def perf_storage_over_nfs_synthetic_tcp_4k(self, result: TestResult) -> None:
...
         self._perf_nfs(result, protocol="udp")
 
+    @TestCaseMetadata(
+        description="""
+        This test case uses fio to test data disk performance with 4K block size.
+        """,
+        priority=3,
+        timeout=TIME_OUT,
+    )
+    def perf_storage_generic_fio_test(self,
```
Comment (dsrivastavv) [inline microsoft/testsuites/performance/storageperf.py]: Can you please explain in comments how these tests are different from `perf_premium_datadisks_4k`?

File: microsoft/testsuites/performance/storageperf.py [hunk 3]
```diff
@@ -187,6 +194,109 @@ def perf_storage_over_nfs_synthetic_tcp_4k(self, result: TestResult) -> None:
...
+            test_name = testcase["testcase_name"]
+            fio_result[test_name] = {}
+            fio_result[test_name]["runtime_param"] = testcase
+
+            try:
+                path = node.get_working_path()
+
+                log.info(f"Executing the FIO testcase : {test_name}")
+
+                name = testcase.get("name", None)
```
Comment (dsrivastavv) [inline microsoft/testsuites/performance/storageperf.py]: These can be handled in a separate class encapsulating these variables

File: microsoft/testsuites/performance/default_fio_testcases.json [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            "direct": 0,
+            "size_mb": 1024,
+            "numjob": 4,
+            "time": 240,
+            "group_reporting": true
+        },
+        {
+            "name": "write",
+            "filename": "Write_Size_4G",
+            "testcase_name": "write_with_4gb",
```
Comment (anirudhrb) [inline microsoft/testsuites/performance/default_fio_testcases.json]: This seems a bit pointless. We can auto generate a name by using the other parameters.

File: microsoft/testsuites/performance/storageperf.py [hunk 5]
```diff
@@ -187,6 +194,109 @@ def perf_storage_over_nfs_synthetic_tcp_4k(self, result: TestResult) -> None:
...
+    def perf_storage_generic_fio_test(self,
+        log: Logger,
+        node: Node,
+        environment: Environment,
+        log_path: Path,
+        result: TestResult,
+        variables: Dict[str, Any],
+    ) -> None:
+        testcases = variables.get("fio_testcase_list", None)
+        if (not testcases):
```
Comment (anirudhrb) [inline microsoft/testsuites/performance/storageperf.py]: redundant parentheses

File: microsoft/testsuites/performance/storageperf.py [hunk 6]
```diff
@@ -187,6 +194,109 @@ def perf_storage_over_nfs_synthetic_tcp_4k(self, result: TestResult) -> None:
     def perf_storage_over_nfs_synthetic_udp_4k(self, result: TestResult) -> None:
         self._perf_nfs(result, protocol="udp")
 
+    @TestCaseMetadata(
+        description="""
+        This test case uses fio to test data disk performance with 4K block size.
```
Comment (anirudhrb) [inline microsoft/testsuites/performance/storageperf.py]: Update description

File: microsoft/testsuites/performance/storageperf.py (hunk 2 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: This test case is not hardcoded and run the test case as per runbook parameters. Even mode for running FIO test case is flexible and can be configured with parameters. This will provide flexibility to run FIO tests easily by just passing parameters in runbook as per need.

Whereas perf_premium_datadisks_4k seems doing same thing but with hardcoded parameters. It will launch FIO test for all 4 modes as perf_disk function run for 4 modes. Please let me know your thoughts on this.

File: microsoft/testsuites/performance/storageperf.py (hunk 3 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: Sure we can do this. I will look for this.

File: microsoft/testsuites/performance/storageperf.py (hunk 1 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: Sure Chi. I will do it.

File: microsoft/testsuites/performance/default_fio_testcases.json [hunk 7]
```diff
@@ -0,0 +1,116 @@
+{
```
Comment (anirudhrb) [inline microsoft/testsuites/performance/default_fio_testcases.json]: We can use the logic of perf_premium_datadisks_4k as the default behavior instead of having this json file. Since perf_premium_datadisks_4k is already in use for Azure VM testing, I would say it is a very good default. This would also be in line with your point that the new test case is generic/configurable for version of perf_premium_datadisks_4k. In the absence of said configuration, it can fallback to perf_premium_datadisks_4k.

File: microsoft/testsuites/performance/storageperf.py (hunk 1 above)
Comment (squirrelsc) [reply microsoft/testsuites/performance/storageperf.py]: I just realized it's another fio test. Please work with Lili to share code as much as possible.

File: microsoft/testsuites/performance/storageperf.py [hunk 8]
```diff
@@ -187,6 +194,109 @@ def perf_storage_over_nfs_synthetic_tcp_4k(self, result: TestResult) -> None:
...
+        node: Node,
+        environment: Environment,
+        log_path: Path,
+        result: TestResult,
+        variables: Dict[str, Any],
+    ) -> None:
+        testcases = variables.get("fio_testcase_list", None)
+        if (not testcases):
+            testcases_file_path = os.path.join(
+                os.path.abspath(__file__).replace(os.path.basename(__file__), ""),
```
Comment (squirrelsc) [inline microsoft/testsuites/performance/storageperf.py]: `os.path.abspath(__file__).parent` is better. If it doesn't work try with pathlib. pathlib is shorter and intuitive.

Comment (squirrelsc) [reply microsoft/testsuites/performance/storageperf.py]: BTW, I feel it doesn't need a file. It can be put in code as dicts.

File: microsoft/testsuites/performance/default_fio_testcases.json (hunk 7 above)
Comment (squirrelsc) [reply microsoft/testsuites/performance/default_fio_testcases.json]: That's fair. There are several things,

1. data driven test cases. If the parameters are only different things like this case. It's definitely should be data driven to share logics as much as possible. The shared data can be dict, or json, it doesn't matter.
2. For hard coded test cases names like `perf_premium_datadisks_4k`. It's necessary. When we test on different scenarios, we need standardized parameters to compare across multiple runs. It shouldn't need to ask every runner to choose from parameters.
3. For customizable parameters like this, it may be needed to handle occasional scenarios. But it shouldn't be overused to create incomparable results.

File: microsoft/testsuites/performance/storageperf.py (hunk 2 above)
Comment (dsrivastavv) [reply microsoft/testsuites/performance/storageperf.py]: How about we modify `perf_premium_datadisks_4k` to work with current hardcoded setting if no parameters are specified, otherwise run it based on parameters? Also, does your testcase work with block size other than `4k`?

Comment (dsrivastavv) [reply microsoft/testsuites/performance/storageperf.py]: My concern is primarily to share as much code as possible. Please feel free to suggest any alternatives

Comment (anirudhrb) [reply microsoft/testsuites/performance/storageperf.py]: `perf_premium_datadisks_4k` has Azure specific naming as well as `requirements`. That's why we can't use it directly.
@smit-gardhariya Why not re-use the underlying `perf_disk` function? If no parameters are provided we can fallback to the configuration used by `perf_premium_datadisks_4k`.

Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: Sure @anirudhrb we can do it. I will make that change

File: microsoft/testsuites/performance/storageperf.py (hunk 1 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: leveraging use of existing perf_disk function

File: microsoft/testsuites/performance/storageperf.py (hunk 2 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: Using perf_disk for all operation having leverage of passing params from runbook as well to run FIO with different configurations. perf_disk is running all FIO mode with all these configs if passed or perf_disk will be executed with default config.

File: microsoft/testsuites/performance/storageperf.py (hunk 5 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: Removed

File: microsoft/testsuites/performance/storageperf.py (hunk 6 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: Done.

File: microsoft/testsuites/performance/default_fio_testcases.json (hunk 7 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/default_fio_testcases.json]: To bring consistency in result, we are launching all test cases using perf_disk only now. If no param passed via runbook, we can launch some default params. This testcase is neither attached to specific scenario nor having specific requirements to run with.

File: microsoft/testsuites/performance/storageperf.py (hunk 8 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: I used file earlier  so we can make add/change testcase without code change. But We are using perf_disk function now with default parameters. This is resolved now.

File: microsoft/testsuites/performance/default_fio_testcases.json (hunk 4 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/default_fio_testcases.json]: It is resolved with new changes as we not have these fields now in runtime parameters

File: microsoft/testsuites/performance/storageperf.py (hunk 3 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: We have used perf_disk now to run testcase with default and runtime parameters. Please check

File: microsoft/testsuites/performance/storageperf.py [hunk 9]
```diff
@@ -210,92 +206,68 @@ def perf_storage_generic_fio_test(self,
...
+                        test_result=result,
+                        test_name=test_name,
+                        num_jobs=num_jobs,
+                        block_size=block_size,
+                        time=time,
+                        size_mb=size_mb,
+                        overwrite=overwrite,
+                    )
+                except Exception as e:
+                    log.error(e)
```
Comment (pupacha) [inline microsoft/testsuites/performance/storageperf.py]: We should let the exception be raised. Else the testcase would get into PASSED state.

File: microsoft/testsuites/performance/storageperf.py [hunk 10]
```diff
@@ -187,6 +189,86 @@ def perf_storage_over_nfs_synthetic_tcp_4k(self, result: TestResult) -> None:
...
+        self,
+        log: Logger,
+        node: Node,
+        environment: Environment,
+        log_path: Path,
+        result: TestResult,
+        variables: Dict[str, Any],
+    ) -> None:
+        testcases = variables.get("fio_testcase_list", None)
+        if not testcases or len(testcases) == 0:
```
Comment (pupacha) [inline microsoft/testsuites/performance/storageperf.py:228]: Instead of calling `per_disk` in both if & else, can we just fill the `testcases` variable with default values in `if` part and let the code run generic without `else` part.

File: microsoft/testsuites/performance/storageperf.py (hunk 9 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: We need try block as we have list of testcases. I have put assert with latest commit to reflect correct status of lisa testcase.

File: microsoft/testsuites/performance/storageperf.py (hunk 10 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py:228]: resolved.

**Review by pupacha — DISMISSED** [review]

File: microsoft/testsuites/performance/common.py [hunk 11]
```diff
@@ -40,11 +38,11 @@ def perf_disk(
...
     filename: str,
-    core_count: int,
-    disk_count: int,
-    disk_setup_type: DiskSetupType,
-    disk_type: DiskType,
     test_result: TestResult,
+    core_count: int = 0,
+    disk_count: int = 0,
+    disk_setup_type: Any = "",
+    disk_type: Any = "",
```
Comment (anirudhrb) [inline microsoft/testsuites/performance/common.py]: Why the change in data types and ordering of the parameters?

Comment (smit-gardhariya) [reply microsoft/testsuites/performance/common.py]: Datatype were specific to already implemented testcases. This is not needed by our generic testcase of FIO and these args are getting used to stamp with notifier message only.  So, We have to make it default to use the same function with our testcase.

Order has to be changed as default args cannot be written before mandatory args.

Comment (anirudhrb) [reply microsoft/testsuites/performance/common.py]: Why not use `Optional` instead of changing type to `Any`?

File: microsoft/testsuites/performance/common.py [hunk 12]
```diff
@@ -40,10 +40,10 @@ def perf_disk(
     start_iodepth: int,
     max_iodepth: int,
     filename: str,
-    core_count: int,
-    disk_count: int,
-    disk_setup_type: DiskSetupType,
-    disk_type: DiskType,
+    core_count: Optional[int],
```
Comment (squirrelsc) [inline microsoft/testsuites/performance/common.py]: What's the reason to mark them optional? In the new change, they are assigned.

File: microsoft/testsuites/performance/storageperf.py [hunk 13]
```diff
@@ -187,6 +189,86 @@ def perf_storage_over_nfs_synthetic_tcp_4k(self, result: TestResult) -> None:
...
+                    size_mb=size_mb,
+                    overwrite=overwrite,
+                    core_count=core_count,
+                    disk_count=1,
+                    disk_setup_type=None,
+                    disk_type=None,
+                )
+            except Exception as err:
+                failed_test_cases.append(testcase)
+                log.error(err)
```
Comment (squirrelsc) [inline microsoft/testsuites/performance/storageperf.py]: It doesn't need to log error message on test case level.

File: microsoft/testsuites/performance/common.py (hunk 11 above)
Comment (squirrelsc) [reply microsoft/testsuites/performance/common.py]: They are metrics, which are collected to database. So it's better to fill them in to get data more accurate.

File: microsoft/testsuites/performance/storageperf.py [hunk 14]
```diff
@@ -187,6 +189,86 @@ def perf_storage_over_nfs_synthetic_tcp_4k(self, result: TestResult) -> None:
...
         self._perf_nfs(result, protocol="udp")
 
+    @TestCaseMetadata(
+        description="""
+        This test case uses fio to test data disk performance.
+        """,
+        priority=3,
+        timeout=TIME_OUT,
+    )
+    def perf_storage_generic_fio_test(
```
Comment (LiliDeng) [inline microsoft/testsuites/performance/storageperf.py:199]: I didn't see any data disk requirement set in case requirement.

File: microsoft/testsuites/performance/storageperf.py (hunk 13 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py]: Sure Chi, i will remove it.

File: microsoft/testsuites/performance/storageperf.py (hunk 14 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/storageperf.py:199]: Yes, this is not specific to any disk type or network interface.

File: microsoft/testsuites/performance/common.py (hunk 12 above)
Comment (smit-gardhariya) [reply microsoft/testsuites/performance/common.py]: Yes in latest push we assigned core_count as per core and disk_count as 1 by default. We need to make disk_setup_type and disk_type as option so we can pass None to them and there will be minimal change in signature of perf_disk.

New testcase does not require disk specification. This is generic FIO test which can run as per passed params in runbook.

I will remove optional from core_count  and disk_count and will keep optional with disk_setup_type and disk_type otherwise flak8 check will fail.

**Comment by smit-gardhariya** [comment]
As discussed with Lili, i have created new PR for better commit traceability.  Closing this PR.

New PR : https://github.com/microsoft/lisa/pull/2325

**Comment by squirrelsc** [comment]
> As discussed with Lili, i have created new PR for better commit traceability. Closing this PR.
> 
> New PR : #2325

It doesn't need to create another PR to merge commits. The new PR lost all comments, it's hard to track changes.
